auto-generated rust code fails to compile

<!-- Thanks for opening an issue! 🤗 -->
<!-- IMPORTANT: This issue tracker is for tracking bugs and features request. For questions or help on how to use Slint, please go to the "Discussions" tab -->
<!-- Please mention your platform and the programming language you are using Slint with -->
<!-- For bugs, please give steps on how to reproduce. What is the expected behavior and what do you see instead. -->
<!-- If possible, please include relevant code snippets, in code blocks (```  ```)-->

### system
- macbook air M1 2020
- macOS 14.5 Sonoma
- `cargo --version` -> `cargo 1.81.0 (2dbb1af80 2024-08-20)`
- `rustup --version` -> `rustc 1.81.0 (eeb90cda1 2024-09-04)`
- `slint = "1.8.0"`

### code
#### build.rs
```rust
fn main() {
    slint_build::compile("ui/app-window.slint").expect("Slint build failed");
}
```

#### src/main.rs
```rust
// Prevent console window in addition to Slint window in Windows release builds when, e.g., starting the app via file manager. Ignored on other platforms.
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

slint::include_modules!();

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let ui = AppWindow::new()?;

    ui.run()?;

    Ok(())
}
```

#### ui/app-window.slint
```slint
import {  HorizontalBox } from "std-widgets.slint";

component DeltasList {
    in property <string> delta;
}

component Changelog inherits Rectangle {
    in-out property <int> count-files-changed;
    in-out property <[string]> changes_log: [];
    in-out property <int> selected: -2;
    callback selection-changed(/*new selection*/ int);
}

export component AppWindow inherits Window {
    in-out property <[string]> changes_log <=> log.changes_log;
    in-out property <int> count-files-changed <=> log.count-files-changed;
    in-out property <string> delta <=> deltas.delta;
    callback request-change-delta(string);
    callback request-diff-delta(string);

    function fill-diff-view() {
        if log.selected >= 0 && log.selected < changes_log.length {
            root.request-change-delta(changes_log[log.selected])
        } else if log.selected == -1 {
            root.request-diff-delta("channel name")
        } else {
            delta = "";
        }
    }

    HorizontalBox {
        log := Changelog {
            horizontal-stretch: 2;
            selection-changed => {
                fill-diff-view();
            }
        }

        deltas := DeltasList {
            horizontal-stretch: 3;
        }
    }
}
```

### problem
running `cargo build` generates the error:
```
error[E0425]: cannot find value `tmp_root_1_changes_log` in this scope
   --> /.../target/debug/build/project-3091f44e836850bc/out/app-window.rs:711:66
    |
711 | ... (match & r#tmp_root_1_changes_log . clone () {
    |              ^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `tmp_log_3_selected` in this scope
   --> /.../target/debug/build/project-3091f44e836850bc/out/app-window.rs:713:47
    |
713 | ...   let index = (r#tmp_log_3_selected . clone ()) as usize ;
    |                    ^^^^^^^^^^^^^^^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `project` (bin "project") due to 2 previous errors
```

this error seems to take place in the generated rust code.

### working alternatives
- replace
    ```slint
    root.request-change-delta(changes_log[log.selected])
    ```
    with
    ```slint
    root.request-change-delta("")
    ```
- replace
    ```slint
    if log.selected >= 0 && log.selected < changes_log.length {
        root.request-change-delta(changes_log[log.selected])
    } else if log.selected == -1 {
        root.request-diff-delta("channel name")
    } else {
        delta = "";
    }
    ```
    with
    ```slint
    if log.selected >= 0 && log.selected < changes_log.length {
        root.request-change-delta(changes_log[log.selected])
    } else {
        root.request-diff-delta("channel name")
    }
    ```

### expected behavior
either:
- compile, or
- emit a slint-contexted error referencing the slint user code, instead of a rust-contexted error referencing the auto-generated rust code

--- a/internal/compiler/passes/deduplicate_property_read.rs
+++ b/internal/compiler/passes/deduplicate_property_read.rs
@@ -21,9 +21,14 @@ pub fn deduplicate_property_read(component: &Component) {
     });
 }
 
+struct ReadCount {
+    count: usize,
+    has_been_mapped: bool,
+}
+
 #[derive(Default)]
 struct PropertyReadCounts {
-    counts: HashMap<NamedReference, usize>,
+    counts: HashMap<NamedReference, ReadCount>,
     /// If at least one element of the map has duplicates
     has_duplicate: bool,
     /// if there is an assignment of a property we currently disable this optimization
@@ -48,12 +53,12 @@ impl<'a> DedupPropState<'a> {
             .counts
             .entry(nr.clone())
             .and_modify(|c| {
-                if *c == 1 {
+                if c.count == 1 {
                     *has_duplicate = true;
                 }
-                *c += 1
+                c.count += 1;
             })
-            .or_insert(1);
+            .or_insert(ReadCount { count: 1, has_been_mapped: false });
     }
 
     fn add_from_children(&self, nr: &NamedReference) -> bool {
@@ -63,10 +68,10 @@ impl<'a> DedupPropState<'a> {
         let mut use_counts = self.counts.borrow_mut();
         let use_counts = &mut *use_counts;
         if let Some(c) = use_counts.counts.get_mut(nr) {
-            if *c == 1 {
+            if c.count == 1 {
                 use_counts.has_duplicate = true;
             }
-            *c += 1;
+            c.count += 1;
             true
         } else {
             false
@@ -75,13 +80,16 @@ impl<'a> DedupPropState<'a> {
 
     fn get_mapping(&self, nr: &NamedReference) -> Option<SmolStr> {
         self.parent_state.and_then(|pr| pr.get_mapping(nr)).or_else(|| {
-            if self.counts.borrow().counts.get(nr).map_or(false, |c| *c > 1) {
-                Some(format_smolstr!("tmp_{}_{}", nr.element().borrow().id, nr.name()))
-            } else {
-                None
-            }
+            self.counts.borrow_mut().counts.get_mut(nr).filter(|c| c.count > 1).map(|c| {
+                c.has_been_mapped = true;
+                map_nr(nr)
+            })
         })
     }
+}
+
+fn map_nr(nr: &NamedReference) -> SmolStr {
+    format_smolstr!("tmp_{}_{}", nr.element().borrow().id, nr.name())
 }
 
 fn process_expression(expr: &mut Expression, old_state: &DedupPropState) {
@@ -93,16 +101,16 @@ fn process_expression(expr: &mut Expression, old_state: &DedupPropState) {
     process_conditional_expressions(expr, &new_state);
     if new_state.counts.borrow().has_set {
         old_state.counts.borrow_mut().has_set = true;
-        return;
+    } else {
+        do_replacements(expr, &new_state);
     }
-    do_replacements(expr, &new_state);
+
     if new_state.counts.borrow().has_duplicate {
         let mut stores = vec![];
-        for (nr, count) in &new_state.counts.borrow().counts {
-            if *count > 1 {
-                let new_name = new_state.get_mapping(nr).unwrap();
+        for (nr, c) in &new_state.counts.borrow().counts {
+            if c.has_been_mapped {
                 stores.push(Expression::StoreLocalVariable {
-                    name: new_name,
+                    name: map_nr(nr),
                     value: Box::new(Expression::PropertyReference(nr.clone())),
                 });
             }
@@ -114,6 +122,9 @@ fn process_expression(expr: &mut Expression, old_state: &DedupPropState) {
 
 // Collect all use of variable and their count, only in non conditional expression
 fn collect_unconditional_read_count(expr: &Expression, result: &DedupPropState) {
+    if result.counts.borrow().has_set {
+        return;
+    }
     match expr {
         Expression::PropertyReference(nr) => {
             result.add(nr);

--- a/tests/cases/issues/issue_6616_optimized_property_read.slint
+++ b/tests/cases/issues/issue_6616_optimized_property_read.slint
@@ -1,0 +1,80 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+
+
+component DeltasList {
+    in property <string> delta;
+}
+
+component Changelog inherits Rectangle {
+    in-out property <int> count-files-changed;
+    in-out property <[string]> changes_log: [];
+    in-out property <int> selected: -2;
+    callback selection-changed(/*new selection*/ int);
+}
+
+export component TestCase inherits Window {
+    in-out property <[string]> changes_log <=> log.changes_log;
+    in-out property <int> count-files-changed <=> log.count-files-changed;
+    in-out property <string> delta <=> deltas.delta;
+    callback request-change-delta(string);
+    callback request-diff-delta(string);
+
+    function fill-diff-view() {
+        if log.selected >= 0 && log.selected < changes_log.length {
+            root.request-change-delta(changes_log[log.selected])
+        } else if log.selected == -1 {
+            root.request-diff-delta("channel name")
+        } else {
+            delta = "";
+        }
+    }
+
+    HorizontalLayout {
+        log := Changelog {
+            selection-changed => {
+                fill-diff-view();
+            }
+        }
+
+        deltas := DeltasList {
+            horizontal-stretch: 3;
+        }
+    }
+
+    //----
+
+    in-out property <int> abc : 25;
+    in-out property <int> new-value: 45;
+    function test-abc() {
+        if new-value != 45 && abc > 0 && abc < 10 {
+            debug("XXX", abc);
+        } else if abc > 45 || new-value == 8 {
+            debug(abc)
+        } else {
+            new-value = abc
+        }
+    }
+    init => { test-abc(); }
+    out property <bool> test: new-value == 25;
+}
+
+/*
+```rust
+let instance = TestCase::new().unwrap();
+assert!(instance.get_test());
+```
+
+```cpp
+auto handle = TestCase::create();
+const TestCase &instance = *handle;
+assert(instance.get_test());
+```
+
+```js
+let instance = new slint.TestCase({});
+assert(instance.test);
+```
+
+*/

